src/v/cluster/feature_manager.cc

```diff
@@ -286,7 +289,65 @@ ss::future<> feature_manager::maybe_log_license_check_info() {
     }
 }
 
+bool feature_manager::need_to_verify_enterprise_license() {
+    return features::is_major_version_upgrade(
+      _feature_table.local().get_active_version(),
+      _feature_table.local().get_latest_logical_version());
+}
+
+void feature_manager::verify_enterprise_license() {
+    vlog(clusterlog.info, "Verifying enterprise license...");
+
+    if (!need_to_verify_enterprise_license()) {
+        vlog(clusterlog.info, "Enterprise license verification skipped...");
+        _verified_enterprise_license = true;
+        return;
+    }
+
+    const auto& license = _feature_table.local().get_license();
+    auto license_missing_or_expired = !license || license->is_expired();
+    auto enterprise_features = report_enterprise_features();
+
+    vlog(
+      clusterlog.info,
+      "Verifying enterprise license: active_version={}, latest_version={}, "
+      "enterprise_features=[{}], license_missing_or_expired={}",
+      _feature_table.local().get_active_version(),
+      _feature_table.local().get_latest_logical_version(),
+      enterprise_features.enabled(),
+      license_missing_or_expired);
+
+    if (enterprise_features.any() && license_missing_or_expired) {
+        throw std::runtime_error{fmt::format(
+          "Looks like you’ve enabled Redpanda Enterprise features ([{}]) "
+          "without a valid license. You cannot upgrade the cluster with "
+          "Redpanda Enterprise features without a valid license. Please "
+          "downgrade this broker to the pre-upgrade version and enter an "
+          "active Redpanda license key (e.g. rpk cluster license set <key>) "
+          "before performing the upgrade. If you don’t have one, please "
+          "request a new/trial license at https://redpanda.com/license-request",
+          enterprise_features.enabled())};
+    }
+
+    _verified_enterprise_license = true;
+}
+
 ss::future<> feature_manager::maybe_update_feature_table() {
```

Comment:
I think I know the answer, but is "running on the backend_shard" a precondition for this function?

Reply:
Yes, there's only a single `feature_manager` running on shard 0, so everything runs there.